src/libraries/System.Text.Json/tests/Serialization/ReferenceHandlerTests.Deserialize.cs

```diff
@@ -1256,7 +1256,7 @@ public static void IdIsNotFirstProperty()
             Assert.Equal("$.$id", ex.Path);
         }
 
-        [Fact]
+        [Fact(Skip = "TODO: Decide whether update the test to no longer point to $id or try to append $id back to the JSON Path.")]
```

Comment:
Before, `AddReferenceOnDeserialize` behave as a TryAdd method, where you could know if the new object was successfully added or not afterwards and we could throw in the call site instead of throwing in the method. 

Now, `AddReference` is a void method that does not follow a TryAdd pattern so we should throw in its implementation, where the code is unaware of the state.

There are a few workaround to keep the behavior as it was before, e.g, we could set `state.Current.JsonPropertyName = "$id"` before calling `AddReference` and set it back to its original value after, but that seems quite ugly and no worth doing without discussion. That's why I left this comment here.

Reply:
Why would you have to set it back to the original value? `Current` is for the specific frame, and we are doing a forward-only read.

Reply:
As we discussed offline, there is one case where we are off of `$id` and we call `AddReference`: for a preserved array the last metadata we will read before calling `AddReference` would be `$values`, i.e. the reader is at the `StartArray` token when we call `AddReference` on the `IEnumerableDefaultConverter`. 

As you suggested, we could potentially avoid the "hacky" thing of remember the last metadata if we call `AddReference` right when we read the `$id` property instead of waiting for `$values`. This probably requires splitting `HandleMetadata` implementation to instantiate the collection when we call `HandleMetadata` from `IEnumerableDefaultconverter`.

I have created issue https://github.com/dotnet/runtime/issues/37168 to follow up on this.